File: 2020/1/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	parsedInput := []int{}
	for _, s := range input {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		parsedInput = append(parsedInput, n)
	}
	fmt.Println(part1(parsedInput))
	fmt.Println(part2(parsedInput))
}

func part1(input []int) int {
	seen := make(map[int]bool)
	for _, n := range input {
		if seen[2020-n] {
			return n * (2020 - n)
		}
		seen[n] = true
	}
	return -1
}

func part2(input []int) int {
	sort.Sort(sort.IntSlice(input))
	z := len(input)
	for i := 0; i < z-2; i++ {
		j := i + 1
		k := z - 1
		for j < k {
			sum := input[i] + input[j] + input[k]
			if sum == 2020 {
				return input[i] * input[j] * input[k]
			}
			if sum < 2020 {
				j++
			} else {
				k--
			}
		}
	}
	return -1
}

File: 2020/1/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n")
  .map(Number);

input.sort((a, b) => a - b);

const target = 2020;

let a = 0;
let b = input.length - 1;

while (a < b && input[a] + input[b] != target) {
  if (input[a] + input[b] > target) {
    b--;
  } else {
    a++;
  }
}

if (a >= b) {
  console.log("Could not find solution for part 1.");
  return;
}
console.log(input[a] * input[b]);

for (let i = 0; i < input.length - 2; i++) {
  let j = i + 1;
  let k = input.length - 1;
  while (j < k && input[i] + input[j] + input[k] != target) {
    if (input[i] + input[j] + input[k] > target) {
      k--;
    } else {
      j++;
    }
  }
  if (input[i] + input[j] + input[k] === target) {
    console.log(input[i] * input[j] * input[k]);
    return;
  }
}

console.log("Could not find solution for part 2.");
